classes/privacy/provider.php

```diff
@@ -24,19 +24,165 @@
 
 namespace block_dedication\privacy;
 
-use core_privacy\local\metadata\null_provider;
+use context_block;
+use core_privacy\local\metadata\collection;
+use core_privacy\local\request\{approved_contextlist, approved_userlist, contextlist, userlist};
+
 /**
  * Class provider
  * @package block_dedication
  */
-class provider implements null_provider {
+class provider implements
+    \core_privacy\local\metadata\provider,
+    \core_privacy\local\request\plugin\provider,
+    \core_privacy\local\request\core_userlist_provider {
+
+    /**
+     * Returns metadata.
+     *
+     * @param collection $collection The initialised collection to add items to.
+     * @return collection A listing of user data stored through this system.
+     */
+    public static function get_metadata(collection $collection): collection {
+        $collection->add_database_table(
+            'block_dedication',
+            [
+                'userid' => 'privacy:metadata:block_dedication:userid',
+                'courseid' => 'privacy:metadata:block_dedication:courseid',
+                'timespent' => 'privacy:metadata:block_dedication:timespent',
+                'timestart' => 'privacy:metadata:block_dedication:timestart',
+            ],
+            'privacy:metadata:block_dedication'
+        );
+        return $collection;
+    }
+
+    /**
+     * Get the list of users who have data within a context.
+     *
+     * @param userlist $userlist The userlist containing the list of users who have data in this context/plugin combination.
+     */
+    public static function get_users_in_context(userlist $userlist) {
+        $context = $userlist->get_context();
+        if (!$context instanceof context_block) {
+            return;
+        }
+        $sql = "SELECT UNIQUE(userid) FROM {block_dedication}";
+        $userlist->add_from_sql('userid', $sql, []);
+    }
+
+    /**
+     * Delete multiple users within a single context.
+     *
+     * @param approved_userlist $userlist The approved context and user information to delete information for.
+     */
+    public static function delete_data_for_users(approved_userlist $userlist) {
```

Comment:
again here - we expect to get a full list of users, see the example here:
https://github.com/UniversityofCanterbury/moodle-block_tasklist/blob/MOODLE_400_STABLE/classes/privacy/provider.php#L73

Reply:
I'm effectively doing the same thing, aren't I? Just using a different approach - using userid IN $userlist->get_userids()

Reply:
yeah - the main issue is your code ignores it when the context !== context_block, it needs to run at the course level too - not all courses will have the block installed but we still store user data in the caching tables.

Reply:
I see! Will make this correction